POLIO-1695 POLIO-1697 POLIO-1698 POLIO-1699 POLIO-1696 POLIO-1700 Vaccine module upload of pdf documents

mathvdh opened the pull request:
Upload PDF files into the vaccine module for each type of file:

- VRF
- Pre Alert
- Form A
- Destruction Report
- Incident report

Related JIRA tickets : POLIO-1695, POLIO-1697, POLIO-1698, POLIO-1699, POLIO-1696, POLIO-1700

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [x] My migrations file are included
- [x] Are there enough tests

## Changes

Add a custom storage manager to point to another public S3 Bucket if `AWS_PUBLIC_STORAGE_BUCKET_NAME` env var is set (otherwise it uses default storage)
Add "document" FileField on relevant models that uses this storage
Modification of the APIs to handle the files and uploads.
Needed a new Parser class to handle Nested Multipart Forms data which is supported by the standard but not by DRF out of the box.
Modify the frontend to handle these documents in all the screens for the relevant items (See above)

## How to test

Either directly via API but you would need to send data in Form Data (Not JSON) (See the code in the tests) or via the UI in all the Vaccine Stock and Supply Chain screens.

## Print screen / video


https://github.com/user-attachments/assets/0c8a7765-b625-4fd2-abe5-1dd615a0010a

mathvdh committed:
IA-3319 Initial work

mathvdh committed:
IA-3319 Add document field on models

mathvdh committed:
Merge branch 'main' of github.com:BLSQ/iaso into IA-3319-vaccine-module-upload-of-pdf-documents

mathvdh committed:
IA-3319 Initial Commit

mathvdh committed:
IA-3319 Pre alerts documents OK

mathvdh committed:
IA-3319 adding FormA document

mathvdh committed:
IA-3319 Add Destruction Report PDF document

mathvdh committed:
IA-3319 Add PDF document to Incident and fix POST

mathvdh committed:
IA-3319 Add requirements

mathvdh committed:
IA-3319 Typo in requirements

mathvdh committed:
IA-3319 Adding translations

mathvdh committed:
IA-3319 Adding tests

mathvdh committed:
IA-3319 Fix problem with rounds format for serializer , Use default storage for files if no env var found

mathvdh committed:
IA-3319 Fix tests

mathvdh committed:
IA-3319 Fix after first tests on staging

quang-le committed:
Merge branch 'main' into IA-3319-vaccine-module-upload-of-pdf-documents

mathvdh committed:
IA-3319 Fix migrations after merging main

quang-le commented on testdata/test_pdf.pdf:
Maybe gitignore this?

mathvdh replied:
I need it for the tests ...

quang-le commented on plugins/polio/js/src/domains/VaccineModule/SupplyChain/hooks/api/vrf.tsx:
```diff
@@ -190,17 +193,139 @@ const getRoundsForApi = (
+            } else {
+                formData.append(key, value);
+            }
+        });
```
This can probably be extracted since you repeat the exact same logic in both `patch2` and `post2`.
Not blocking

mathvdh replied:
Done

quang-le commented on plugins/polio/js/src/domains/VaccineModule/SupplyChain/hooks/api/vrf.tsx:
```diff
@@ -190,17 +193,139 @@ const getRoundsForApi = (
-    const payload: Partial<VRF> = {
-        ...vrf,
-        rounds: getRoundsForApi(vrf?.rounds),
+    const filteredParams = vrf
```
nitpick: `filteredParams` sounds like you're preparing query params for a GET request. Not blocking

quang-le commented on plugins/polio/js/src/domains/VaccineModule/SupplyChain/Details/VaccineRequestForm/VaccineRequestForm.tsx:
```diff
@@ -15,9 +15,20 @@ import {
+    if (!Array.isArray(err_docs)) return [err_docs]; 
+    else return err_docs;
+};
+
```
I would put this somewhere else, like in utils or something

mathvdh replied:
Done

mathvdh committed:
IA-3319 Fix content types

mathvdh committed:
IA-3319 Comma separated values

quang-le committed:
IA-3319: fixes in front-end

quang-le committed:
IA-3319: fix round error
- Covert round to comma separated string if it's an array
- remove manual setting of header
- pass use FormData as request body in all cases

mathvdh committed:
IA-3319 Cleanup and refactor

mathvdh committed:
IA-3319 Refactoring

bramj commented on plugins/polio/api/vaccines/supply_chain.py:
```diff
@@ -301,10 +310,25 @@ class Meta:
+                data["rounds"] = [{"number": num} for num in rounds]
+            except Exception as e:
+                raise serializers.ValidationError(f"Invalid rounds data: {e}")
+        return super().to_internal_value(data)
```
This looks unrelated to the document upload feature, right?

mathvdh replied:
It's only related because we changed from JSON to FormData to send the data to the server @bramj and so the old way of sending the rounds was not working

bramj reviewed (approved):
Tested locally and working! Nothing to comment on the backend code, looks good to me, nice job :+1: 
I only wonder where we are able to see the uploaded documents in the frontend? But maybe that's not important here.

mathvdh committed:
Merge branch 'main' of github.com:BLSQ/iaso into IA-3319-vaccine-module-upload-of-pdf-documents

mathvdh committed:
IA-3319 Fix migrations after merge from main

mathvdh committed:
IA-3319 Fix bad merge
